coc/ext/triggers/triggers.py

```diff
@@ -0,0 +1,472 @@
+                                if isinstance(res, Exception):
+                                    await self.__handle_exception(func, arg, res)
+                        else:
+                            await fixture()
+                    except Exception as e:
+                        await self.__handle_exception(func, None, e)
+
+                    # sleep until next execution time
+                    next_run = self.next_run
+                    if self.logger and datetime.now().astimezone() <= next_run:
+                        self.logger.info(
+                            f'{self.__class__.__name__} finished for {func.__name__}. Next run: {next_run.isoformat()}'
+                        )
+                    elif self.logger:  # i.e. next_run is in the past
+                        self.logger.warning(
+                            f'{self.__class__.__name__} missed the scheduled run time for {func.__name__}. Running now'
+                        )
+
+                    await self.sleep_until(next_run)
+
+            # create a reference to the repeating task to prevent it from accidentally being garbage collected
+            self.task = self.loop.create_task(inner())
+
+        if self.autostart:  # immediately start the trigger
+            if self.logger:
+                self.logger.info(f'{self.__class__.__name__} for {func.__name__} auto-started')
+            self.loop.create_task(wrapped())
+        else:  # add trigger to registry
+            trigger_registry.append(wrapped())
+            if self.logger:
+                self.logger.info(f'{self.__class__.__name__} for {func.__name__} registered for manual start')
+
+        return wrapped
+
+    async def __handle_exception(self, func: CoroFunction, arg: Any, exc: Exception):
+        """Handle exceptions during trigger calls. This will attempt to call the provided logger and
+        any available error handler
+        """
+
+        if self.logger:
+            self.logger.error(f'function: {func.__name__}, failing iter_arg: {arg}\n'
+                              ''.join(format_exception(type(exc), exc, exc.__traceback__)))
+
+        error_handler = self.error_handler or default_error_handler
+        if error_handler:
+            await error_handler(func.__name__, arg, exc)
+
+    @staticmethod
+    async def sleep_until(wakeup_date: datetime):
+        """Sleep until a defined point in time. If that point is in the past, don't sleep at all
+        Parameters
+        ----------
+        wakeup_date: :class:`datetime.datetime`
+            a timezone-aware datetime at which the trigger should wake up again
+        """
+
+        await asyncio.sleep(max((wakeup_date - datetime.now().astimezone()).total_seconds(), 0))
+
+    @property
+    def next_run(self) -> datetime:
+        """Calculate the date and time of the next run. Needs to be overwritten in subclasses"""
+        raise NotImplementedError('All `BaseTrigger` subclasses need to implement `next_run`')
```

Comment:
Why not making this an `@abstractmethod`?

Reply:
Felt like a property was appropriate for this. No real other reason for or against anything, both work equally fine

Reply:
property is absolutely fine. I thought more about stacking both decorators on top of each other.

Reply:
I suppose one could to that, but then it'd be `@staticmethod`(which you probably meant in the first place)

Reply:
Fair point. Added